### ISSUE TITLE
myloader: CPU at 400%, but restore process is very slow

### ISSUE THREAD

**Issue by inkonito** [mydumper/mydumper#557]
I am working on migrating a database (340Go) from mysql 5.7 to mysql 8, both on azure PaaS, by using an ubuntu VM in the middle.
I used mydumper very easily, and got my dump folder done in about 3h and 20minutes, no issues.

I am now trying to load this dump in my mysql 8 database with this command:
`myloader -h xxx.mysql.database.azure.com -u xxx --password='xxx' -d /data/dump --queries-per-transaction=500 --compress-protocol --threads=8 -L /data/myloader.log -v 3 --innodb-optimize-keys`

With the `top` command, i can see that the CPU is at 400% (4vCore), but the network interface is doing almost nothing, and the disks are not doing so much either... Except for the CPU, metrics are showing that not so much is going on. On the DB side, not so much also, after 6+ hours, i had only 2Go of data in the new database. I stopped the process and tried others options, reducing/increasing number of threads, etc. with no luck. 
The process is not stuck, since there are some logs sometimes saying that it's processing a new file. But still, i feel something is not right here. I stopped after 6h, and was only on file 35 on 725 (!).

Any help on what is going on or what i am doing wrong ?

**Comment by davidducos** [comment]
Hi @inkonito , Can you share the log file? What version are you using? can you share also a `show processlist` and a `show engine innodb status\G` when the process is working?

**Comment by inkonito** [comment]
Hello @davidducos 

I'm using `myloader 0.11.5, built against MySQL 5.7.33-36` but my destination DB is MySQL 8.0.21, maybe that's an issue ?

You'll find the log file, and the output of both `show processlist` and `show engine innodb status\G` [in this gist](https://gist.github.com/inkonito/3581bbe2679c3b658083cea1e080e82b). 
This was captured during a process with 32 threads, that seemed to accelerate the migration, but there is still something very strange...

Thank you for your help, and let me know if you need something else.

**Comment by davidducos** [comment]
Yes, I need the logs and yes it is slow. Looks like it is not continuing with the files in the queue.

**Comment by inkonito** [comment]
New logs are [in this gist](https://gist.github.com/inkonito/0b3e845659896abcb9c59163af1ef519)
The process seems not to be stuck, new files are processed, but it is very slow yes.

Thanks again for your help anyway

**Comment by davidducos** [comment]
In the show process list, I can see the myloader connections in sleep... that shouldn't be happening. Are all innodb tables? can you try with --max-threads-per-table=1 and let me know if the threads stay in sleep state?

**Comment by inkonito** [comment]
Every table is an innodb table, yes. 

I tried with --max-threads-per-table=1, and the threads are almost always in sleep state, but sometimes, there is one thread that is doing an insert, for example: 

```
| 27 | myuser          | 10.230.40.6:50428 | mydb | Query   |    1 | update                 | INSERT INTO `vehicle_management_manual_transaction_staging_errors` VALUES(1,69602),(1,69603),(3276 |
```
But again, that's only one of the 32 threads, and most of the time, they are all sleeping.

This time, it seems to be a bit faster at first, after 5minutes, 100 of the 649 files are already done, but after that it seems to slow down again... I'll wait a little to see if it speeds up or not.

**Comment by davidducos** [comment]
Can you execute ls -l on the backup directory?

**Comment by inkonito** [comment]
Sure, [here you are in this Gist](https://gist.github.com/inkonito/5d3780a7c2d4f51955a483786bdc4ac8)

**Comment by inkonito** [comment]
Hi @davidducos 
I did some investigations with custom build and by adding logs in the myloader.c file.
All my threads ends up hanging on this line: [myloader.c#L1700](https://github.com/mydumper/mydumper/blob/a8df9b16afea7e64d04e3370d1b63dfc2854d700/myloader.c#L1700)
Can you help me understand why? Is it because i did the dump with files too big ? I used these options during mydumper process: `--chunk-filesize=1024 --statement-size=104857600` 
What can i do to avoid that ?

**Comment by inkonito** [comment]
OK thank you. What do you suggest ? That i wait for a fix, or to restart my dump with another value for statement-size ? What value do you suggest ?

**Comment by davidducos** [comment]
@inkonito , I found that setting --statement-size up to 4MB works fine, above that number myloader start failing. Then I realize that it is related to max_allowed_packet in on the server. I recommend you to set max_allowed_package to a size higher than the value that you use in --statement-size.

**Comment by davidducos** [comment]
We need to add a section as [myloader_variables] like [myloader_global] to set global variables that are going to be reverted after myloader has been executed. This will allow us to keep a standard --defaults-file where we can recommend users to set the best setups.

**Comment by inkonito** [comment]
OK, thank you again. I'll try with 4mb tomorrow. But i already set the max allowed pack age on the db side to be 1go, and from the logs i got, the issue was not happening when communication with the DB.
Anyway, thank you for the recommendations

**Comment by inkonito** [comment]
Hello @davidducos 
I tried making my dump with 4Mb statement size, but the import was still slow.
So i continued digging more, and found that [this commit](https://github.com/mydumper/mydumper/commit/486047453b170381da06a9d919596410bc8ce789) has a huge impact on performance when it comes to big files (i have many sql file that > 100Mo, even > 1Go)
I commented out the code regarding the line tracking on error, and everything seems to run smoothly now...

**Comment by davidducos** [comment]
Oh shit... you are totally right... let me see how to improve that

**Comment by inkonito** [comment]
Hi @davidducos .
I tried your branch, and everything seems to run smoothly. After 2 hours of running, i have passed 70% of the files for a 342Go dump.
Nice work, and thank you again for your help !

**Comment by R-omk** [comment]
@davidducos 

myloader can stuck    and utilize 100% per mydumer thread 

![image](https://github.com/mydumper/mydumper/assets/1633879/1e087106-eec5-4d0f-80ac-de24e173b5b2)

```
myloader  --version
myloader v0.16.1-3, built against MySQL 8.0.36-28 with SSL support

```


`show processlist;     `   show's that all process sleep

command 
```
myloader --directory  export-20240407-155730  -h XXX.XXX.XXX.XXX -u root   -v 3  -p 'XXXXXXXX'  --max-threads-per-table 3 -t 9  --logfile /tmp/mydumper7.log --disable-redo-log --innodb-optimize-keys AFTER_IMPORT_PER_TABLE
```

last line in logs:

```
[INFO] - Fast index creation will be use for table: XXXXX.XXXXX

```

it  stuck  and table XXXXX.XXXXX  was not created

**Comment by R-omk** [comment]
related https://github.com/mydumper/mydumper/issues/1372#issuecomment-2017458002

similar symptoms, also not all threads work at the same time

**Comment by davidducos** [comment]
Hi @R-omk,
Can you build a test case? btw, did you modify the metadata file or removed files from the backup dir?

You should be creating a new bug for your specific scenario.
